### PR TITLE
fix: align APP_PAGE rscData reads with Next FileSystemCache (fixes #102)

### LINKED ISSUE #102: Postponed/fallback RSC cache entries cause stale router.refresh and Connection closed errors (1.18.0)
## Summary

In the published **1.18.0** package, `RedisStringsHandler.get()` returns cached `rscData` for fallback and postponed PPR entries. These buffers can be incomplete Flight streams. Next's filesystem cache deliberately omits them in these cases, but the Redis handler returns them unchanged.

In a production-mode Next.js application, this produced stale UI after native `revalidateTag(..., { expire: 0 })` followed by `router.refresh()`, occasional locale mixing, and React error 412 (decoded from the bundled RSC client as **“Connection closed.”**). A full document reload recovered. Direct requests to each server already contained the updated data.

I checked **PR #95 and the released 1.18.0 npm artifact**, rather than assuming this was still present from 1.17.0. The attached contract reproduction still fails on unmodified 1.18.0 and passes with the focused patch below.

## Environment

- `@trieb.work/nextjs-turbo-redis-cache`: 1.18.0 (also observed on 1.17.0)
- Next.js: 16.3.5
- `redis`: 4.7.0
- Valkey: 9.0, cluster mode
- Node.js: 24.11.1
- App configuration: `cacheComponents: true`, `partialPrefetching: true`, shared `cacheHandler` and `cacheHandlers`, `cacheMaxMemorySize: 0`

The failing read contract does not depend on cross-slot invalidation; this is separate from #101 and #97.

## Expected vs actual behavior

The pinned [Next.js FileSystemCache.get implementation](https://github.com/vercel/next.js/blob/v16.3.5/packages/next/src/server/lib/incremental-cache/file-system-cache.ts) only loads the full RSC file when:

```ts
!ctx.isFallback && (!ctx.isRoutePPREnabled || meta?.postponed == null)
```

The Redis handler should return an equivalent entry. Instead, it retains `rscData` for every `APP_PAGE` entry, including incomplete fallback/postponed payloads. Next's app-page runtime can then take the static-Flight response path, and the client reaches the end of the stream with unresolved data.

## Reproduction and patch

The standalone files are included below as expandable code blocks:

- `repro.cjs`: comparison against the real pinned Next filesystem cache, using actual Valkey reads/writes.
- `postponed-rsc-1.18.0.patch`: focused patch for the published package, including TypeScript, CommonJS/ESM bundles, and declarations.

Start a disposable local Valkey cluster-mode primary:

```sh
docker run --rm -d --name valkey-rsc-repro \
  -p 127.0.0.1:6398:6379 valkey/valkey:9.0 \
  valkey-server --cluster-enabled yes \
  --cluster-config-file /tmp/nodes.conf \
  --notify-keyspace-events Exe --save "" --appendonly no
until docker exec valkey-rsc-repro valkey-cli ping; do sleep 1; done
docker exec valkey-rsc-repro valkey-cli CLUSTER ADDSLOTSRANGE 0 16383
until docker exec valkey-rsc-repro valkey-cli CLUSTER INFO | grep -q 'cluster_state:ok'; do sleep 1; done
```

Save the two files below in an empty scratch directory, then run:

```sh
npm init -y
npm install --save-exact @trieb.work/nextjs-turbo-redis-cache@1.18.0 \
  redis@4.7.0 next@16.3.5 react@19.2.0 react-dom@19.2.0
CACHE_TEST_REDIS_URL=redis://127.0.0.1:6398 node repro.cjs
```

The unpatched package fails the first comparison:

```text
AssertionError [ERR_ASSERTION]: postponed PPR: rscData
actual: Buffer(...)
expected: undefined
```

After applying only this patch, all **16 checks pass**: eight cases across CommonJS and ESM. Cases cover nonempty/empty postponed state, fallback reads, complete PPR/non-PPR pages, and preservation of the stored payload after a fallback read.

The proposed read normalization is:

```ts
const value = cacheEntry.value as SetCacheValue;
if (
  ctx.kind === 'APP_PAGE' &&
  value?.kind === 'APP_PAGE' &&
  (ctx.isFallback || (ctx.isRoutePPREnabled && value.postponed != null))
) {
  return { ...cacheEntry, value: { ...value, rscData: undefined } };
}
```

This leaves cached HTML, resume state, headers, and segment prefetch buffers intact. It normalizes the returned entry without deleting or mutating the stored payload, and it also handles existing warm entries. The source type currently spells the field `postboned`; the patch corrects it to `postponed`.

## Additional integration evidence

The original browser failure was reproduced locally with the 1.17.0-based integration in a full production/Turbopack app: three independent web processes behind a round-robin proxy, shared Valkey, a local data API, and a local realtime broadcaster. Adding this read normalization made four viewers across two locales follow three consecutive data changes without reloading or changing locale; the final document reload also passed.

The same three-instance full-app check also passed after upgrading to 1.18.0 with this RSC correction and the remaining cursor/incremental cluster fixes.

Both sides of that full-app comparison retained the same separate Next.js cold-fallback compatibility patch. The attached library-level reproduction requires **no Next.js patch** and independently confirms this defect against the released 1.18.0 handler.

Could the incremental handler adopt the filesystem cache's RSC read contract and add coverage for postponed and fallback entries?

## Standalone files

The reproduction intentionally exits 1 on the unpatched package. Apply the patch and repeat:

```sh
cd node_modules/@trieb.work/nextjs-turbo-redis-cache
git apply ../../../postponed-rsc-1.18.0.patch
cd ../../..
CACHE_TEST_REDIS_URL=redis://127.0.0.1:6398 node repro.cjs
```

The reproduction only deletes its own uniquely prefixed keys. Stop the disposable server with `docker stop valkey-rsc-repro` when finished.

<details>
<summary>postponed-rsc-1.18.0.patch</summary>

```diff
diff --git a/dist/index.d.mts b/dist/index.d.mts
--- a/dist/index.d.mts
+++ b/dist/index.d.mts
@@ -93,7 +93,7 @@
     html: string;
     rscData: Buffer;
     segmentData: unknown;
-    postboned: unknown;
+    postponed: unknown;
 } | {
     kind: 'APP_ROUTE';
     status: number;
diff --git a/dist/index.d.ts b/dist/index.d.ts
--- a/dist/index.d.ts
+++ b/dist/index.d.ts
@@ -93,7 +93,7 @@
     html: string;
     rscData: Buffer;
     segmentData: unknown;
-    postboned: unknown;
+    postponed: unknown;
 } | {
     kind: 'APP_ROUTE';
     status: number;
diff --git a/dist/index.js b/dist/index.js
--- a/dist/index.js
+++ b/dist/index.js
@@ -916,6 +919,10 @@
           }
         }
       }
+      const value = cacheEntry.value;
+      if (ctx.kind === "APP_PAGE" && value?.kind === "APP_PAGE" && (ctx.isFallback || ctx.isRoutePPREnabled && value.postponed != null)) {
+        return { ...cacheEntry, value: { ...value, rscData: void 0 } };
+      }
       return cacheEntry;
     } catch (error) {
       console.error(
diff --git a/dist/index.mjs b/dist/index.mjs
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -872,6 +875,10 @@
           }
         }
       }
+      const value = cacheEntry.value;
+      if (ctx.kind === "APP_PAGE" && value?.kind === "APP_PAGE" && (ctx.isFallback || ctx.isRoutePPREnabled && value.postponed != null)) {
+        return { ...cacheEntry, value: { ...value, rscData: void 0 } };
+      }
       return cacheEntry;
     } catch (error) {
       console.error(
diff --git a/src/RedisStringsHandler.ts b/src/RedisStringsHandler.ts
--- a/src/RedisStringsHandler.ts
+++ b/src/RedisStringsHandler.ts
@@ -58,7 +58,7 @@
       html: string;
       rscData: Buffer;
       segmentData: unknown;
-      postboned: unknown;
+      postponed: unknown;
     }
   | {
       kind: 'APP_ROUTE';
@@ -660,6 +660,20 @@
         }
       }
 
+      // Match Next's FileSystemCache: a fallback or postponed PPR entry does
+      // not contain a complete Flight response. Returning its partial rscData
+      // makes Next serve an unfinished stream as a full navigation response.
+      // Keep the HTML, postponed state and segment prefetch data available so
+      // Next can resume rendering through its normal request path.
+      const value = cacheEntry.value as SetCacheValue;
+      if (
+        ctx.kind === 'APP_PAGE' &&
+        value?.kind === 'APP_PAGE' &&
+        (ctx.isFallback || (ctx.isRoutePPREnabled && value.postponed != null))
+      ) {
+        return { ...cacheEntry, value: { ...value, rscData: undefined } };
+      }
+
       return cacheEntry;
     } catch (error) {
       // This catch block is necessary to handle any errors that may occur during:
```

</details>

<details>
<summary>repro.cjs</summary>

```js
/* Compare the Redis handler with the pinned Next.js filesystem cache.
 * A postponed/fallback Flight stream must not be served as a complete response.
 * Requires the local Compose Valkey cluster. No production services are used.
 */
const assert = require("node:assert/strict");
const fs = require("node:fs/promises");
const os = require("node:os");
const path = require("node:path");
const { fork } = require("node:child_process");
const { pathToFileURL } = require("node:url");
const { createClient } = require("redis");
const FileSystemCache =
  require("next/dist/server/lib/incremental-cache/file-system-cache").default;
const {
  RSC_SEGMENTS_DIR_SUFFIX,
  RSC_SEGMENT_SUFFIX,
} = require("next/dist/lib/constants");

async function child() {
  const library = process.env.RSC_PACKAGE_FILE.endsWith(".mjs")
    ? await import(pathToFileURL(process.env.RSC_PACKAGE_FILE).href)
    : require(process.env.RSC_PACKAGE_FILE);
  const redis = new library.RedisStringsHandler({
    redisUrl: process.env.REDIS_URL,
    database: 0,
    keyPrefix: process.env.RSC_KEY_PREFIX,
  });
  const disk = new FileSystemCache({
    fs,
    serverDistDir: path.join(process.env.RSC_FIXTURE_DIR, "server"),
    maxMemoryCacheSize: 0,
    flushToDisk: true,
    revalidatedTags: [],
  });
  const cases = [
    {
      name: "postponed PPR",
      ppr: true,
      fallback: false,
      postponed: "resume-state",
    },
    {
      name: "empty postponed state",
      ppr: true,
      fallback: false,
      postponed: "",
    },
    {
      name: "PPR fallback",
      ppr: true,
      fallback: true,
      postponed: "resume-state",
    },
    { name: "fallback without postponed state", ppr: true, fallback: true },
    { name: "legacy fallback", ppr: false, fallback: true },
    { name: "complete PPR", ppr: true, fallback: false },
    {
      name: "complete PPR with null state",
      ppr: true,
      fallback: false,
      postponed: null,
    },
    { name: "complete non-PPR", ppr: false, fallback: false },
  ];
  for (const [i, test] of cases.entries()) {
    const key = `/page-${i}`;
    const value = {
      kind: "APP_PAGE",
      html: "<h1>Cached event</h1>",
      rscData: Buffer.from('0:{"event":"updated"}\n'),
      postponed: test.postponed,
      segmentData: new Map([["/_tree", Buffer.from("prefetch tree")]]),
      headers: {
        "x-next-cache-tags": "rsc-regression",
        "x-nextjs-stale-time": "300",
      },
      status: 200,
    };
    // Seed the artifacts Next's builder writes, including the full .rsc file.
    const base = path.join(
      process.env.RSC_FIXTURE_DIR,
      "server/app",
      key.slice(1),
    );
    await fs.mkdir(base + RSC_SEGMENTS_DIR_SUFFIX, { recursive: true });
    await fs.writeFile(base + ".html", value.html);
    await fs.writeFile(base + ".rsc", value.rscData);
    await fs.writeFile(
      base + ".meta",
      JSON.stringify({
        headers: value.headers,
        status: value.status,
        postponed: value.postponed,
        segmentPaths: ["/_tree"],
      }),
    );
    await fs.writeFile(
      base + RSC_SEGMENTS_DIR_SUFFIX + "/_tree" + RSC_SEGMENT_SUFFIX,
      value.segmentData.get("/_tree"),
    );
    const context = {
      kind: "APP_PAGE",
      isFallback: test.fallback,
      isRoutePPREnabled: test.ppr,
    };
    await redis.set(key, value, {
      ...context,
      cacheControl: { revalidate: 3600 },
    });
    const expected = await disk.get(key, context);
    const actual = await redis.get(key, context);
    assert.ok(
      expected && actual,
      `${test.name}: both caches must contain the page`,
    );
    for (const field of [
      "rscData",
      "html",
      "postponed",
      "segmentData",
      "headers",
      "status",
    ]) {
      assert.deepEqual(
        actual.value[field],
        expected.value[field],
        `${test.name}: ${field}`,
      );
    }
    // A fallback read must not mutate the stored value or deduplicated read.
    const completeContext = {
      ...context,
      isFallback: false,
      isRoutePPREnabled: false,
    };
    const complete = await redis.get(key, completeContext);
    assert.deepEqual(
      complete.value.rscData,
      value.rscData,
      `${test.name}: preserve stored Flight data`,
    );
    console.log(`PASS ${test.name}: matches Next FileSystemCache`);
  }
}

async function main() {
  const url = process.env.CACHE_TEST_REDIS_URL || "redis://127.0.0.1:6381";
  assert.ok(
    ["localhost", "127.0.0.1", "[::1]"].includes(new URL(url).hostname),
    "Only local Valkey is allowed",
  );
  const client = createClient({ url });
  await client.connect();
  const root = await fs.mkdtemp(path.join(os.tmpdir(), "cache-rsc-"));
  const prefix = `rsc-regression:${process.pid}:${Date.now()}:`;
  try {
    assert.match(await client.info("server"), /valkey_version:9\./);
    assert.match(
      await client.sendCommand(["CLUSTER", "INFO"]),
      /cluster_state:ok/,
    );
    const dist = process.env.CACHE_RSC_PACKAGE_DIR
      ? path.join(process.env.CACHE_RSC_PACKAGE_DIR, "dist")
      : path.dirname(require.resolve("@trieb.work/nextjs-turbo-redis-cache"));
    for (const format of ["js", "mjs"]) {
      const worker = fork(__filename, ["--child"], {
        env: {
          ...process.env,
          REDIS_URL: url,
          SKIP_KEYSPACE_CONFIG_CHECK: "true",
          RSC_PACKAGE_FILE: path.join(dist, `index.${format}`),
          RSC_KEY_PREFIX: prefix + format + ":",
          RSC_FIXTURE_DIR: path.join(root, format),
        },
        stdio: ["ignore", "inherit", "inherit", "ipc"],
      });
      await new Promise((resolve, reject) => {
        const timeout = setTimeout(() => {
          worker.kill();
          reject(new Error(`${format}: timed out`));
        }, 15000);
        worker.once("error", (error) => {
          clearTimeout(timeout);
          reject(error);
        });
        worker.once("exit", (code) => {
          clearTimeout(timeout);
          code === 0
            ? resolve()
            : reject(new Error(`${format}: regression failed`));
        });
      });
      console.log(`${format}: all 8 cache contract cases passed`);
    }
  } finally {
    for await (const key of client.scanIterator({ MATCH: prefix + "*" }))
      await client.unlink(key);
    await client.quit();
    await fs.rm(root, { recursive: true, force: true });
  }
}
if (process.argv.includes("--child")) {
  child().then(
    () => process.exit(0),
    (error) => {
      console.error(error);
      process.exit(1);
    },
  );
} else {
  main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
  });
}
```

</details>

### PR DESCRIPTION
**Pull request by JannikZed**
## Summary

- Normalize `RedisStringsHandler.get()` for `APP_PAGE` so fallback and postponed PPR entries do not return partial `rscData`, matching Next.js 16.3 `FileSystemCache`.
- Rename type field `postboned` → `postponed`; still honor legacy `postboned` in stored entries.
- Add contract tests against Next `FileSystemCache` plus edge cases (legacy field, deduped reads, non-APP_PAGE kinds).

Fixes #102

## Test plan

- [x] `pnpm vitest --run test/vitest/unit/postponed-rsc-read.contract.test.ts` (12 tests)
- [x] `pnpm test:unit`

### CODE REVIEW

File: test/vitest/unit/postponed-rsc-read.contract.test.ts [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+  };
+  if (!version.startsWith('16.3.')) {
+    throw new Error(
+      `Next.js contract tests require 16.3.x from the fixture app, got ${version} at ${nextPkgJson}`,
+    );
+  }
+  return createRequire(nextPkgJson);
+}
+
+const requireNext = loadNextFixtureRequire();
```
[the remark this answers is not in the document]
Comment (JannikZed) [reply test/vitest/unit/postponed-rsc-read.contract.test.ts]: Fixed in 0acb270. Fixture resolution no longer throws at module load: `resolveNextFixture()` returns `null` when the Next.js 16.3 fixture is missing (or the wrong version), and the `FileSystemCache`-backed suite is now wrapped in `describe.runIf(!!nextFixture)`. The mock-only edge-case suite stays unconditional. Verified both paths: without the fixture, `pnpm test:unit` runs all 13 files (91 passed, 10 skipped, no abort); with the fixture installed it runs the full contract (100 passed, 1 skipped). The setup command is also documented in test/README.md.

File: test/vitest/unit/postponed-rsc-read.contract.test.ts [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+import { createRequire } from 'node:module';
+import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
+
+import RedisStringsHandler from '../../../src/RedisStringsHandler';
+
+const NEXT_FIXTURE_APP = path.join(
+  __dirname,
+  '../../nextjs-test-projects/next-app-16-3-0',
+);
+const nextPkgJson = path.join(
```
[the remark this answers is not in the document]
Comment (JannikZed) [reply test/vitest/unit/postponed-rsc-read.contract.test.ts]: Addressed via the "narrow and document the versions covered" option you suggested (0acb270). I audited Next's FileSystemCache across the supported range: 16.1, 16.2 and 16.3 share an identical rscData read condition (`!ctx.isFallback && (!ctx.isRoutePPREnabled || meta?.postponed == null)`), while 15.0.3–16.0.x use a separate `.rsc.prefetch` code path and never read the full `.rsc` file for PPR reads. So no version in the supported range serves a postponed PPR entry's full `.rsc` payload through rscData — the normalized read is aligned with every supported release, and the contract is pinned against the 16.3.0 fixture, which is representative of the 16.1–16.3 behavior. The rationale is documented in the test header and test/README.md. Adding 15.x/16.0 fixtures would exercise a different code path the fix doesn't target, so documenting the covered scope is the more accurate coverage here.
